src/main/java/com/github/maxopoly/finale/combat/AsyncPacketHandler.java

```diff
@@ -0,0 +1,76 @@
+package com.github.maxopoly.finale.combat;
+
+import org.bukkit.World;
+import org.bukkit.entity.Damageable;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Player;
+
+import com.comphenix.protocol.PacketType;
+import com.comphenix.protocol.events.ListenerPriority;
+import com.comphenix.protocol.events.PacketAdapter;
+import com.comphenix.protocol.events.PacketContainer;
+import com.comphenix.protocol.events.PacketEvent;
+import com.comphenix.protocol.reflect.StructureModifier;
+import com.comphenix.protocol.wrappers.EnumWrappers.EntityUseAction;
+import com.github.maxopoly.finale.Finale;
+
+import net.md_5.bungee.api.ChatColor;
+
+public class AsyncPacketHandler extends PacketAdapter {
+	
+	private double maxReach;
+	private int cpsLimit;
+	
+	public AsyncPacketHandler() {
```

Comment:
The thinking behind it is that by listening and processing the attack packet asynchronously with no delay then synchronizing the hits, you get more responsive combat as you don't have to wait for the processing of the attack packets every tick; they've already been processed so it's just the emptying of a queue of hits each tick. Obviously, you get the same outcome in essence, but this is something a lot of big PvP servers do, so I assume while it seems minimal, there are benefits to reap during large fights.

Reply:
You don't have to wait for their processing anyway, Minecraft will process all available packets every tick. I don't know what big servers do, but at least this implementation offers no benefits compared to vanilla handling and is entirely unnecessary imo. It's functionality should just be moved into bukkit events.

Reply:
I just realized, even if there is no benefits to handling it asynchronously; it is still necessary to handle this ourselves because of the modifications to the NMS attack method.